library/firewall_lib_facts.py

```diff
@@ -0,0 +1,220 @@
+
+        settings = fetch_settings_from_dir(default_setting_dir, detailed, fw)
+        if settings:
+            defaults[setting] = settings
+        settings = fetch_settings_from_dir(custom_setting_dir, True, fw)
+        if settings:
+            custom[setting] = settings
+
+    if defaults:
+        config["default"] = defaults
+    if custom:
+        config["custom"] = custom
+
+    return config
+
+
+def fetch_settings_from_dir(directory, detailed=False, fw=None):
+    setting_options = [
+        _file[:-4] for _file in os.listdir(directory) if _file.endswith(".xml")
+    ]
+    if not detailed:
+        return setting_options
+    else:
+        setting_name = os.path.basename(directory)
+        settings = []
+        for _item in setting_options:
+            element_settings = {}
+            if setting_name == "zones":
+                element = fw.config().getZoneByName(_item).getSettings()
+                try:
+                    element_settings = element.getSettingsDict()
+                except AttributeError:
+                    element_settings["version"] = element.getVersion()
+                    element_settings["short"] = element.getShort()
+                    element_settings["description"] = element.getDescription()
+                    element_settings["target"] = element.getTarget()
+                    element_settings["services"] = element.getServices()
+                    element_settings["ports"] = element.getPorts()
+                    element_settings["icmp_blocks"] = element.getIcmpBlocks()
+                    element_settings["masqerade"] = element.getMasquerade()
+                    element_settings["forward_ports"] = element.getForwardPorts()
+                    element_settings["interfaces"] = element.getInterfaces()
+                    element_settings["sources"] = element.getSources()
+                    element_settings["rules_str"] = element.getRichRules()
+                    element_settings["protocols"] = element.getProtocols()
+                    element_settings["source_ports"] = element.getSourcePorts()
+                    element_settings[
+                        "icmp_block_inversion"
+                    ] = element.getIcmpBlockInversion()
+            elif setting_name == "services":
+                element = fw.config().getServiceByName(_item).getSettings()
+                try:
+                    element_settings = element.getSettingsDict()
+                except AttributeError:
+                    element_settings["version"] = element.getVersion()
+                    element_settings["short"] = element.getShort()
+                    element_settings["description"] = element.getDescription()
+                    element_settings["protocols"] = element.getProtocols()
+                    element_settings["source_ports"] = element.getSourcePorts()
+                    element_settings["helpers"] = element.getModules()
```

Comment:
This actually fetches modules, so I'll change this to `element_settings["modules"]` to also better represent the alternative dictionary `getSettingsDict()` provides.